### PR TITLE
fix(dot/state): actually prune finalized tries from memory

### PR DESCRIPTION
**Pull request by noot**
## Changes

- fix removal of tries older than current finalized from memory
- exchange `BlockAnnounceHandshake` with peer on connect, previously it would sometimes take many minutes for the node to start bootstrap syncing. this makes it start basically as soon as we get peers.

## Tests

```
run gossamer for many hours and see better memory usage (hopefully)?
```

## Issues

- #1973 

## Primary Reviewer

- @qdm12

### CODE REVIEW

**Comment by noot** [comment]
on this brach: 
- at block 24k and around 20min uptime: the memory usage was 11.2% of 15.4gb = 1.725gb
- at bloock 213k and ~3h uptime: 52.6% of 15.4gb = 8.1gb

on development: 
- at block 26k around 22min uptime: 12.9% of 15.5gb = 2gb
= at block 183k and ~3h uptime: 60.4% of 15.5gb = 9.362gb

seems like there's moderate improvement with these changes, but memory usage still slowly creeps up. will continue to run the nodes and report back

**Comment by noot** [comment]
on this branch at 3h36min, block height 250k, usage = 63% of 15.4gb = 9.7gb
on development at 3h36min, block height 215k, usage = 76.8% of 15.5gb = 11.9gb

it seems after the 3 hour mark is when the memory usage really starts to increase, seems quite a bit better here

File: dot/network/service.go [hunk 1]
```diff
@@ -430,6 +430,25 @@ func (s *Service) sentBlockIntervalTelemetry() {
 func (s *Service) handleConn(conn libp2pnetwork.Conn) {
 	// TODO: currently we only have one set so setID is 0, change this once we have more set in peerSet.
 	s.host.cm.peerSetHandler.Incoming(0, conn.RemotePeer())
+
+	// exchange BlockAnnounceHandshake with peer so we can start to
+	// sync if necessary.
+	prtl, has := s.notificationsProtocols[BlockAnnounceMsgType]
```
Comment (qdm12) [inline dot/network/service.go:436]: Why not `protocol` :cry:

File: dot/network/service.go [hunk 2]
```diff
@@ -430,6 +430,25 @@ func (s *Service) sentBlockIntervalTelemetry() {
...
+	if !has {
+		return
+	}
+
+	hs, err := prtl.getHandshake()
+	if err != nil {
+		return
```
Comment (qdm12) [inline dot/network/service.go:443]: Maybe log it out? :thinking: Does this happen often?

Same question for `sendHandshake` below

File: dot/state/block_finalisation.go [hunk 3]
```diff
@@ -235,8 +235,17 @@ func (bs *BlockState) handleFinalisedBlock(curr common.Hash) error {
...
+		block, has = bs.getAndDeleteUnfinalisedBlock(hash)
+		if !has {
+			continue
+		}
+
+		logger.Tracef("cleaned out finalised block from memory; block number %s with hash %s", block.Header.Number, hash)
+
+		go func(header *types.Header) {
+			bs.pruneKeyCh <- header
+		}(&block.Header)
```
Comment (qdm12) [inline dot/state/block_finalisation.go:248]: Do we really need to run this in a goroutine? Can this be blocking :thinking:

File: lib/blocktree/blocktree.go [hunk 4]
```diff
@@ -133,20 +133,20 @@ func (bt *BlockTree) getNode(h Hash) (ret *node) {
 // Prune sets the given hash as the new blocktree root,
 // removing all nodes that are not the new root node or its descendant
 // It returns an array of hashes that have been pruned
-func (bt *BlockTree) Prune(finalised Hash) (pruned []Hash) {
+func (bt *BlockTree) Prune(finalised Hash) []Hash {
```
Comment (qdm12) [inline lib/blocktree/blocktree.go:136]: nit What's wrong with the named return? :smiling_face_with_tear:

**Review by qdm12 — APPROVED** [review]
Improvements :100: ! Although not the thick silver bullet I have been dreaming of lately :smile:

File: lib/blocktree/blocktree.go (hunk 4 above)
Comment (noot) [reply lib/blocktree/blocktree.go]: nothing, i was just trying things out lol, changed it back

File: dot/state/block_finalisation.go (hunk 3 above)
Comment (noot) [reply dot/state/block_finalisation.go:248]: no it shouldn't be blocking anymore, will update

File: dot/network/service.go (hunk 2 above)
Comment (noot) [reply dot/network/service.go:443]: i don't think this would ever really happen, I can add a log though

File: dot/network/service.go (hunk 1 above)
Comment (noot) [reply dot/network/service.go:436]: there's a package already imported at `protocol`, i can change it though if you want!

Comment (qdm12) [reply dot/network/service.go:436]: Ah yes, one of the dumb Go feature/bug 😢 

Maybe have it as `proto`? Or rename the import to `protocolpkg` 🤔 Up to you!

File: dot/network/service.go (hunk 2 above)
Comment (qdm12) [reply dot/network/service.go:447]: Yeah log it out at the error level so we can see if it happens? If it does happen a lot, we can just then either handle it or log it at the trace level.

File: dot/network/service.go [hunk 5]
```diff
@@ -440,15 +440,23 @@ func (s *Service) handleConn(conn libp2pnetwork.Conn) {
...
+			prtl.protocolID,
+			err,
+		)
 		return
 	}
 
-	stream, err := s.sendHandshake(conn.RemotePeer(), hs, prtl)
+	_, err = s.sendHandshake(conn.RemotePeer(), hs, prtl)
 	if err != nil {
+		logger.Debugf("failed to send handshake to peer on connection, peer=%s: %s",
```
Comment (qdm12) [inline dot/network/service.go:452]: supernit making it more sentence oriented

```suggestion
		logger.Debugf("failed to send handshake to peer %s on connection: %s",
```

**Review by qdm12 — APPROVED** [review]

File: dot/network/service.go [hunk 6]
```diff
@@ -440,15 +440,23 @@ func (s *Service) handleConn(conn libp2pnetwork.Conn) {
...
+			conn.RemotePeer(),
+			err,
+		)
 		return
 	}
 
-	_ = stream.Close()
+	// leave stream open if there's no error
```
[the remark this answers is not in the document]
Comment (noot) [reply dot/network/service.go:459]: yeah, `sendHandshake` will close the stream if it returns and error

File: dot/network/service_test.go [hunk 7]
```diff
@@ -100,7 +100,7 @@ func createTestService(t *testing.T, cfg *Config) (srvc *Service) {
 			Port:         availablePort(t),
 			NoBootstrap:  true,
 			NoMDNS:       true,
-			LogLvl:       4,
+			LogLvl:       5,
```
[the remark this answers is not in the document]
Comment (noot) [reply dot/network/service_test.go:103]: yes will do!

File: dot/network/service.go (hunk 5 above)
Comment (noot) [reply dot/network/service.go]: done

**Review by qdm12 — CHANGES_REQUESTED** [review]

File: dot/state/block_finalisation.go (hunk 3 above)
Comment (noot) [reply dot/state/block_finalisation.go]: woops, i updated the other part but not here. done!

**Review by qdm12 — APPROVED** [review]

**Review by EclesioMeloJunior — APPROVED** [review]

**Pull request merged**